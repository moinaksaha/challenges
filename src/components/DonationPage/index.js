// @flow
import React, { Component } from 'react';
import PropTypes from 'prop-types';

import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { withRouter } from 'react-router-dom';
import styled from 'styled-components';
import Sticky from 'react-stickynode';
// import loadable from '@loadable/component';

import { fetchAllCharities } from '../../store/actions/charities';
import { fetchAllDonations } from '../../store/actions/donations';

import { summaryDonationMultiCurrency } from '../../utils/helpers';

import Card from '../Card';
import Header from '../Header';
import GenericLoader from '../GenericLoader';
import GenericError from '../GenericError';
import GenericNoData from '../GenericNoData';
// const Card = loadable(() => import('../Card'));
// const Header = loadable(() => import('../Header'));

const AllCharities = styled.div`
  display: flex;
  flex-wrap: wrap;
  z-index: 1;
  margin-top: 20px;
`;

export class DonationPage extends Component {

  constructor(props) {
    super(props);
    this.state = {
      isHeaderSticky: false,
    }
  }
  componentDidMount() {
    const { fetchAllCharities, fetchAllDonations } = this.props;
    fetchAllCharities();
    fetchAllDonations();
  }

  getCharityDonation = (charityID) => {
    const { allDonations } = this.props;
    if (allDonations && allDonations.data && allDonations.data.length !== 0) {
      return allDonations.data.filter((item) => item.charitiesId === charityID)
    }
    return [];
  }

  handleStateChange = ({status}) => {
    this.setState({
      isHeaderSticky: (status === 2) ? true : false,
    })
  }

  render() {
    const { allCharities, allDonations } = this.props;
    const { isHeaderSticky } = this.state;

    if (allCharities.loading) {
      return <GenericLoader />;
    }

    if (allCharities.error) {
      return <GenericError />;
    }
    if (!allCharities || !allCharities.data || allCharities.data.length === 0) {
      return <GenericNoData />
    }

    // Only works if all donations are made in the same currency
    // Improved that logic

    // const totalDonation = (allDonations && allDonations.data && allDonations.data.length !== 0) ?
    //   summaryDonations(allDonations.data.map((item) => (item.amount))):
    //   null;

    const totalDonation = (allDonations && allDonations.data && allDonations.data.length !== 0) ?
      summaryDonationMultiCurrency(allDonations.data):
      null;

    return (
      <div>
        <Sticky enabled={true} top={0} innerZ={9} activeClass={'moinak'} onStateChange={this.handleStateChange}>
          <Header totalDonation={totalDonation} isStuck={isHeaderSticky}/>
        </Sticky>

        <AllCharities>
          {allCharities.data.map((item) => {
            return (
              <Card 
                key={item.id}
                data={item}
                donations={this.getCharityDonation(item.id)}
              />
            )
          })}
        </AllCharities>
      </div>
    );
  }
};

DonationPage.propTypes = {
  fetchAllCharities: PropTypes.func,
  fetchAllDonations: PropTypes.func,
  allCharities: PropTypes.object,
  allDonations: PropTypes.object,
};

const mapStateToProps = state => {
  return {
    allCharities: state.charities.charities,
    allDonations: state.donations.donations,
  };
};
    
const mapDispatchToProps = dispatch =>
  bindActionCreators(
    {
      fetchAllCharities,
      fetchAllDonations,
    },
    dispatch
  );
    
export default withRouter(
  connect(
    mapStateToProps,
    mapDispatchToProps
  )(DonationPage)
);


