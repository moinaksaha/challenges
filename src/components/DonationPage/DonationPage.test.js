import React from 'react'
import DonationPage from './index.js'
// import Card from '../Card'
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
// import { BrowserRouter as Router } from 'react-router-dom';
// import sinon from 'sinon';

// import { Provider } from 'react-redux';
// import configureMockStore from 'redux-mock-store';

describe('Component: <DonationPage />', () => {
  
  test('Render Status: Success', () => {
    const wrapper = shallow(<DonationPage />);
    expect(wrapper.exists()).toBe(true);
  })

  test('Snapshot: Enzyme', () => {
    const tree = shallow(<DonationPage />)
    expect(toJson(tree)).toMatchSnapshot()
  })


  // const mockCharities = {
  //   data: [
  //     {
  //       id: 1,
  //       name: 'Baan Kru Noi',
  //       image: 'baan-kru-noi.jpg',
  //       currency: 'THB',
  //     },
  //     {
  //       id: 2,
  //       name: 'Habitat for Humanity Thailand',
  //       image: 'habitat-for-humanity-thailand.jpg',
  //       currency: 'THB',
  //     },
  //     {
  //       id: 3,
  //       name: 'Paper Ranger',
  //       image: 'paper-ranger.jpg',
  //       currency: 'THB',
  //     },
  //     {
  //       id: 4,
  //       name: 'Makhampom Theater',
  //       image: 'makhampom-theater.jpg',
  //       currency: 'THB',
  //     },
  //     {
  //       id: 5,
  //       name: 'Thailand Association of the Blind',
  //       image: 'thailand-association-of-the-blind.jpg',
  //       currency: 'THB',
  //     },
  //   ],
  //   loading: false,
  //   error: null,
  // }

  // test('Calls componentWillReceiveProps', () => {
  //   // const spy = sinon.spy(DonationPage.prototype, 'componentWillReceiveProps');
  //   const mockStore = configureMockStore();
  //   const store = mockStore({});

  //   const wrapper = shallow(
  //     <Provider store={store}>
  //       <Router>
  //         <DonationPage allCharities={mockCharities} />
  //       </Router>
  //     </Provider>);
  //   // const wrapper = mount(<DonationPage allCharities={mockCharities} />);

  //   expect(wrapper.prop('allCharities')).toEqual(mockCharities)
  //   // expect(spy).to.have.property('callCount', 0);
  //   // wrapper.setProps({ allCharities: mockCharities });
  //   // expect(spy).to.have.property('callCount', 1);
  // })

  

  // // const initialState = {
  // //   charities: {
  // //     loading: false,
  // //     data: null,
  // //     error: null,
  // //   },
  // // }
  
  // const initialState = {
  //   fetchAllDonations() {},
  //   fetchAllCharities() {},
  //   charities: {
  //     loading: false,
  //     data: null,
  //     error: null,
  //   },
  //   donations: [],
  // }

  // test('Renders 5 instances of card for the given mock data', () => {
  //   const mockStore = configureMockStore();
  //   const store = mockStore(initialState);

  //   const wrapper = shallow(
  //     <Provider store={store}>
  //       <Router>
  //         <DonationPage />
  //       </Router>
  //     </Provider>);
  //   wrapper.setProps({ allCharities: mockCharities });
  //   expect(wrapper.find(Card)).to.have.lengthOf(5);
  // })
  
})