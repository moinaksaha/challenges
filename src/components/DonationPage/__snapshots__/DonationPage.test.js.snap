// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: <DonationPage /> Snapshot: Enzyme 1`] = `
<Route>
  <Component />
</Route>
`;
